### PR TITLE
Priority-based Push and Pull

### PR DESCRIPTION
**Pull request by anandj91**
Messages sent based on an assigned priority
Higher priority messages gets preference over lower priority ones

### CODE REVIEW

**Comment by anandj91** [comment]
This is an additional feature ps-lite has to support in order to implement priority-based parameter propagation in MXNet (JIRA ticket - https://issues.apache.org/jira/browse/MXNET-1294).

File: include/ps/internal/customer.h [hunk 1]
```diff
@@ -12,7 +12,7 @@
 #include <thread>
 #include <memory>
 #include "ps/internal/message.h"
-#include "ps/internal/threadsafe_queue.h"
+#include "ps/internal/threadsafe_pqueue.h"
 namespace ps {
```
Comment (eric-haibin-lin) [inline include/ps/internal/customer.h]: Could we also have an MXNet PR that update the 3rdparty dependency to `anandj91:prio` temporarily to verify that all mxnet kvstore test cases pass?

File: src/zmq_van.h [hunk 2]
```diff
@@ -110,6 +111,8 @@ class ZMQVan : public Van {
     if (my_node_.id != Node::kEmpty) {
       std::string my_id = "ps" + std::to_string(my_node_.id);
       zmq_setsockopt(sender, ZMQ_IDENTITY, my_id.data(), my_id.size());
+      int hwm = 1;
```
Comment (eric-haibin-lin) [inline src/zmq_van.h]: Why is this change necessary?

File: include/ps/internal/threadsafe_pqueue.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  };
+  mutable std::mutex mu_;
+  std::priority_queue<Message, std::vector<Message>, Compare> queue_;
+  std::condition_variable cond_;
+};
+
+}  // namespace ps
+
+// bool TryPop(T& value) {
```
Comment (eric-haibin-lin) [inline include/ps/internal/threadsafe_pqueue.h]: please remove unused code

File: src/zmq_van.h (hunk 2 above)
Comment (anandj91) [reply src/zmq_van.h]: This basically makes the amq_send calls blocking. This was necessary because otherwise the messages in the newly added priority queue will get polled immediately and pushed onto the zmq's FIFO queue. This defeats the purpose of adding a priority queue.

File: include/ps/internal/customer.h (hunk 1 above)
Comment (anandj91) [reply include/ps/internal/customer.h]: https://github.com/apache/incubator-mxnet/pull/15124
I have raised a PR on MXNet. There are some test cases failing now. Some of them are failing because of build issues. I'll take a look into it and fix it soon.

File: include/ps/internal/threadsafe_pqueue.h (hunk 3 above)
Comment (anandj91) [reply include/ps/internal/threadsafe_pqueue.h]: done.

**Comment by eric-haibin-lin** [comment]
Could you resolve the conflict?

**Comment by bobzhuyb** [comment]
@anandj91 I noticed that you modified the Van base class, especially the new Push() interface. I want to understand whether this is backward compatible, e.g., with the recently merged RDMA Van? https://github.com/dmlc/ps-lite/pull/151

IMHO, a cleaner implementation is to keep the Van base class, just inherit from it (or from ZMQVan) and create your own Van, e.g., P3Van. Users are free to use it or fall back to other options, by setting environmental variables.

**Comment by bobzhuyb** [comment]
@anandj91 The latest P3Van looks good! Thank you.
